command/snapshot/inspect/snapshot_inspect.go

```diff
@@ -56,27 +65,181 @@ func (c *cmd) Run(args []string) int {
+	readFile, meta, err := snapshot.Read(nil, f)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error reading snapshot: %s", err))
+	}
+	stats, offset, err := enhance(readFile)
 	if err != nil {
-		c.UI.Error(fmt.Sprintf("Error verifying snapshot: %s", err))
+		c.UI.Error(fmt.Sprintf("Error extracting snapshot data: %s", err))
 		return 1
 	}
+	// Outputs the original style of inspect information
+	legacy, err := c.legacyStats(meta)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error outputting snapshot data: %s", err))
+	}
+	c.UI.Info(legacy.String())
+
+	// Outputs the more detailed snapshot information
+	enhanced, err := c.readStats(stats, offset)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error reading snapshot stats: %s", err))
+		return 1
+	}
+	c.UI.Info(enhanced.String())
+
+	return 0
+}
 
+// legacyStats outputs the expected stats from the original snapshot
+// inspect command
+func (c *cmd) legacyStats(meta *raft.SnapshotMeta) (bytes.Buffer, error) {
 	var b bytes.Buffer
 	tw := tabwriter.NewWriter(&b, 0, 2, 6, ' ', 0)
 	fmt.Fprintf(tw, "ID\t%s\n", meta.ID)
 	fmt.Fprintf(tw, "Size\t%d\n", meta.Size)
 	fmt.Fprintf(tw, "Index\t%d\n", meta.Index)
 	fmt.Fprintf(tw, "Term\t%d\n", meta.Term)
 	fmt.Fprintf(tw, "Version\t%d\n", meta.Version)
-	if err = tw.Flush(); err != nil {
+	if err := tw.Flush(); err != nil {
 		c.UI.Error(fmt.Sprintf("Error rendering snapshot info: %s", err))
-		return 1
+		return b, err
 	}
+	return b, nil
+}
+
+type typeStats struct {
+	Name  string
+	Sum   int
+	Count int
+}
 
-	c.UI.Info(b.String())
+// countingReader helps keep track of the bytes we have read
+// when reading snapshots
+type countingReader struct {
+	wrappedReader io.Reader
+	read          int
+}
 
-	return 0
+func (r *countingReader) Read(p []byte) (n int, err error) {
+	n, err = r.wrappedReader.Read(p)
+	if err == nil {
+		r.read += n
+	}
+	return n, err
+}
+
+// enhance utilizes ReadSnapshot to populate the struct with
+// all of the snapshot's itemized data
+func enhance(file io.Reader) (map[structs.MessageType]typeStats, int, error) {
+	stats := make(map[structs.MessageType]typeStats)
+	var offset int
+	cr := &countingReader{wrappedReader: file}
+	offset = 0
+	handler := func(header *fsm.SnapshotHeader, msg structs.MessageType, dec *codec.Decoder) error {
+		name := structs.MessageType.String(msg)
+		s := stats[msg]
+		if s.Name == "" {
+			s.Name = name
+		}
+		var val interface{}
+		err := dec.Decode(&val)
+		if err != nil {
+			return fmt.Errorf("failed to decode msg type %v, error %v", name, err)
+		}
+
+		size := cr.read - offset
+		s.Sum += size
+		s.Count++
+		offset = cr.read
+		stats[msg] = s
+		return nil
+	}
+	if err := fsm.ReadSnapshot(cr, handler); err != nil {
+		return nil, 0, err
+	}
+	return stats, offset, nil
+
+}
+
+// readStats takes the information generated from enhance and creates human
+// readable output from it
+func (c *cmd) readStats(stats map[structs.MessageType]typeStats, offset int) (bytes.Buffer, error) {
```

Comment:
minor: renaming `offset` to `totalSize` might make this variable more clear. It comes from outset, but here it is being used as the total size.